### ISSUE TITLE
Wildcard redirects

### ISSUE THREAD

**Issue by RealOrangeOne** [wagtail/wagtail#11557]
### Is your proposal related to a problem?

Currently, it's not possible to specify redirects with wildcards in them, which allow matching anything under a path.

### Describe the solution you'd like

It should be possibly to specify a wildcard(s) as part of the redirect source URL, and then redirect to a single page. For example, redirecting `/foo/*` to `/bar`.

In an ideal world, it'd be nice to use capture groups to extract parts of the URL and extract them elsewhere, but that might be better as a second phase.

### Describe alternatives you've considered

I implemented part of a solution in https://github.com/wagtail/wagtail/pull/7528, which worked fairly well in theory, but required using `%` as the wildcard placeholder, and leant hard on `MATCHES` in SQL, which may not be great for usability.

### Additional context

Here are some examples of how Cloudflare handles redirects: https://developers.cloudflare.com/rules/url-forwarding/single-redirects/examples/

### Working on this

Anyone can contribute to this. View our [contributing guidelines](https://docs.wagtail.org/en/latest/contributing/index.html), add a comment to the issue once you’re ready to start.

Redirects are checked as part of any page's 404 response, so it's important this process be performant.

**Comment by rohitsrma** [comment]
@RealOrangeOne I would like to give this a try. I'll really appreciate if you could provide guidance on how I can approach implementing this feature.

**Comment by RealOrangeOne** [comment]
I don't have an implementation in mind, you might need to do some research into potential ways to achieve this.

**Comment by rohitsrma** [comment]
Thanks @RealOrangeOne, I believe we can update the _get_redirect method in RedirectMiddleware to handle wildcard redirects.
https://github.com/wagtail/wagtail/blob/e97f3648c76628059da6fd7d7be6d0e102c77de5/wagtail/contrib/redirects/middleware.py#L11
I think this way we can implement the feature with minimal changes. Here's an example of how we could do it:

```
wildcard_redirects = models.Redirect.get_for_site(site).filter(old_path__endswith='*')
    for redirect in wildcard_redirects:
        wildcard_path = redirect.old_path.rstrip('*')
        if path.startswith(wildcard_path):
            return redirect
```
I tried this locally and it is working good. I would love to know your thoughts on this.

**Comment by RealOrangeOne** [comment]
Prefix matching is definitely better than nothing, but proper glob redirects would be nice. You might be able to achieve that by finding the first `*` in a string, doing a DB query for every redirect which starts with that, then running `glob.glob` on those matched paths? Not the most efficient, but it's still probably fairly fast, without needing to add a lot of extra complexity into the DB (using a regex probably isn't what we want)

**Comment by rohitsrma** [comment]
Thanks @RealOrangeOne, I followed your suggestions and created a draft. Can you please share your thoughts on it?

**Comment by RealOrangeOne** [comment]
I've taken a look and added some comments. On second thoughts, it might not be the best approach. Looking for any redirect with a `*` in is going to be quite an expensive operation. I think the only way to do this is in the DB, which does increase the complexity a fair bit.

**Comment by RealOrangeOne** [comment]
That's a good reference, yes. Although I think relying on DB wildcard literals may not be best. Perhaps pre-processing the search paths to use a `*` rather than `%`, but exactly how to do that I don't know. I'm also not sure if there are any security implications around using user-controlled strings in `MATCHES` statements (there probably isn't, but I don't know).

Regexes might also work, but that'd require users entering regexes for redirects, which has its own problems, both in terms of usability and security.

**Comment by rohitsrma** [comment]
Oh okay, Could preprocessing the path potentially introduce unforeseen problems, especially if deliberate use of '*' could lead to unexpected behavior in the database? Also, I'm not aware of the potential security risks associated with this approach. 

Therefore, should we consider sticking with the current method(following your suggestions on it) until we can find a proper way to handle this through the database?

**Comment by RealOrangeOne** [comment]
Wildcard redirects turn out to be both quite difficult to implement, and likely not super useful to many to warrant the added overhead.

Looking at https://pypi.org/project/django-redirects/, another redirects implementation for Django, it supports exact, prefix, or full regex. Supporting regexes is probably more complex than we want, and might be better suited for elsewhere, but prefix ought not to be too complex, and give most of the benefits. Adding a "type" field, so a redirect can be either exact or a prefix, and a simple `__startswith` query won't impact performance too much.

With that said, this is all 1 person's fairly quick thinking about how this ought to work. This probably wants some proper thinking about the _right_ way to implement this, what we're trying to solve, and if it's even something we want.

**Comment by rohitsrma** [comment]
@RealOrangeOne I completely agree with what you said. So, following your suggestion, I am holding off on this until we have proper input and thoughts from other members as well!

**Comment by markc-versa** [comment]
Just as a datapoint, since pattern-based redirects weren't implemented, I just had to roll my own, making things more brittle and less standardized. There are definitely tradeoffs to being too conservative when adding bells and whistles to a platform.

**Comment by lb-** [comment]
https://github.com/wagtail/wagtail/pull/11628 was started as a draft towards fixing this.

However, there were additional considerations (see comment on this issue - https://github.com/wagtail/wagtail/issues/11557#issuecomment-1959897859 ) that needed to be worked through.

Anyone is free to review that (now closed) PR and see if they want to give this another go.

**Comment by Mr0grog** [comment]
FWIW, I recently had a use case for something like this, but need the wildcard part of the original URL to be part of the redirected URL. In my case, a sub-site that lived under `/my-prefix` was moved to a separate domain, so I’d want to redirect `/my-prefix/some/page` to `https://newsite.com/some/page`.

If things worked like django-redirects, I think I’d need something like the regex to make this happen (AFAICT the “prefix” style is only for matching, and doesn’t give you a way to put the bit that matched the wildcard onto the end of the redirected URL).

If you’re [still] looking for syntaxes that might be nicer than a regex, I think Netlify’s approach is not too bad: https://docs.netlify.com/manage/routing/redirects/redirect-options/#splats
- If you have a `*` in the URL to match, you can put `:splat` in the redirected URL to be replaced with whatever matched the wildcard.
- If you want more complex matching, you can have placeholders prefixed with `:`, for example:
    - From `/news/:month/:date/:year/:slug`
    - To `/blog/:year/:month/:date/:slug`

There are also two standards for this kind of thing, but they are pretty complicated, and probably don’t buy much over Regexes other than being purpose-designed for URLs:
- WHATWG `URLPattern` ([spec](https://urlpattern.spec.whatwg.org), [MDN explainer](https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API)).
- [RFC 6570 (URI Templates)](https://datatracker.ietf.org/doc/rfc6570/).
